Changes to facilitate CFA-0.6

davidhassell opened the pull request:
Needs https://github.com/NCAS-CMS/cfdm/pull/223 to be merged before review.

* Introduces "fragment array" classes and updates `NetCDFRead` to deal with aggregation variables; but all of the CFA-specific code is in  `NetCDFRead` is temporarily disabled. 
* Reorganises `cf/data/` to group array classes together.
* Refactors `cf/data/creation.py` to move some "to dask" functionality to `Array` subclasses.
* Made minor changes  to `test_dsg.py` and `test_gathering.py` to check that the new framework works.
* 
The purpose of this PR is to get some CFA placeholders in, but not to use or test any of the CFA code. However, the `cf/data/` rearrangement and `creation.py` changes are used and it would be good to merge these before any further work gets done on those areas.

sadielbartholomew commented:
@davidhassell https://github.com/NCAS-CMS/cfdm/pull/223 was recently merged, so I going to start reviewing this, but there are some merge conflicts and GitHub isn't letting me resolve them myself (which I might have done if they were trivial and non-controversial, just so I could start reviewing immediately). If you can resolve those ASAP then I'll get the review done. Thanks.

davidhassell commented:
Hi Sadie - after a slightly scary command line intervention, I think it's all OK now ...

sadielbartholomew commented:
> I think it's all OK now ...

Thanks David. From a quick eyeball of the changeset here, and running of some tests with the updated branch, it looks like the conflict resolution did indeed go fine. I'll start a proper review now and aim to finish and submit it by the end of the day.

sadielbartholomew commented on cf/data/array/fullarray.py:
```diff
@@ -0,0 +1,185 @@
+
+        :Returns:
+
+            asdasda TODO
```
```suggestion
            TODODASKDOCS
```

sadielbartholomew commented on cf/data/array/fullarray.py:
```diff
@@ -0,0 +1,185 @@
+
+        .. versionadded:: TODODASKVER
+
+        To cfdm.Array ?
```
```suggestion
        TODODASKDOCS
```

unless you can remember where you were going with that thought :smile:

sadielbartholomew commented on cf/data/array/fullarray.py:
```diff
@@ -0,0 +1,185 @@
+                    Number of elements indexin the data array.
+
+                fill_value : scalar, optional
+
+        #        masked_all: `bool`
```
```suggestion
```

sadielbartholomew commented on cf/data/array/fullarray.py:
```diff
@@ -0,0 +1,185 @@
+                dtype : numpy.dtype
+                    The numpy data type of the data array.
+
+            shape : tuple
+                The data array's dimension sizes.
```
I see the `TODODASKDOCS` above but might as well fix the parameters already listed...

```suggestion
                shape : tuple
                    The data array's dimension sizes.
```

sadielbartholomew commented on cf/data/array/cfanetcdfarray.py:
```diff
@@ -0,0 +1,693 @@
+
+    """
+
+    def __new__(cls, *args, **kwargs):
```
+1 for nice use of `__new__`

sadielbartholomew reviewed (approved):
I've raised some minor comments, mostly regarding docstrings, but overall this is all good. Please merge once you've considered all of the minor feedback, and the below.

As part of reviewing I generated some UML diagrams of the overall `data` class and module structure using `pyreverse` to check that everything looked sensible in the 'new way'. It all looks good to me, but thought I'd share here in case you wanted to review them too (they are large images so you'll need to e.g. right click to open in a new tab and then explore by moving about and/or zooming in):

### Classes
![classes_DASK-CFA-DATA](https://user-images.githubusercontent.com/30274190/202278880-17c9f438-397e-491c-8746-5f416ffbf810.png)

### Modules
![packages_DASK-CFA-DATA](https://user-images.githubusercontent.com/30274190/202278884-337a9e6e-32af-4051-94df-5c9a939beec2.png)

davidhassell commented on cf/__init__.py:
```diff
@@ -199,7 +199,7 @@
 
 # Check the version of dask
-_minimum_vn = "2022.6.0"
+_minimum_vn = "2022.1.0"
```
Certainly not safe! I'll be fixing that ....

davidhassell commented on cf/read_write/netcdf/netcdfread.py:
```diff
@@ -401,3 +425,52 @@ def _array_from_variable(self, ncvar):
+
+        :Returns:
+
+            `CFANetCDFArray`, `dict`
```
Compromise? `(CFANetCDFArray, dict)`

davidhassell commented on cf/data/data.py:
```diff
@@ -443,34 +446,117 @@ def __init__(
+        #                "Can't define chunks for dask input arrays. Consider "
+        #                "rechunking the dask array before initialisation, or "
+        #                "rechunking the Data after initialisation."
+        #            )
```
+1

davidhassell commented on cf/data/array/umarray.py:
```diff
@@ -0,0 +1,665 @@
+                ignored.
+
+                The number of uncompressed array dimensions.
+
```
Thanks - https://github.com/NCAS-CMS/cf-python/pull/476/commits/a068d2bb7062616e506947c8617b44a5b7f7b611

davidhassell commented on cf/read_write/netcdf/netcdfread.py:
```diff
@@ -227,9 +207,72 @@ def _create_data(
+
+        cfa_04 = attributes.get("cf_role") == "cfa_variable"
+        if cfa_04:
+            # TODOCFA: Modify this message for v4.0.0.
```
I was clearly confused! They're all `TODOCFA`s, now.

davidhassell commented:
Hi Sadie, thanks for the thorough review - I've addressed all of your points. Those diagrams are excellent! They've got me thinking about a few things (like should `NetCDFFragmentArray` be a subclass of `NetCDFArray`, partly so that it can pick up active storage capability (https://github.com/NCAS-CMS/cf-python/issues/501)?), but we'll revisit  all that when we do the final, full implementation of CFA.
